File: java-gdb-activities-20/activity19-solution/src/com/gdb/logging/FileLogDestination.java
package com.gdb.logging;

import com.gdb.command.TransactionCommand;
import java.io.IOException;
import java.util.List;

public class FileLogDestination implements LogDestination {
    private TransactionLog log;

    public FileLogDestination() {
        this.log = new TransactionLog();
    }

    @Override
    public void write(TransactionCommand cmd) {
        try {
            log.log(cmd);
        } catch (IOException e) {
            throw new RuntimeException("Error writing to file log: " + e.getMessage(), e);
        }
    }

    @Override
    public List<TransactionCommand> readAll() {
        try {
            return log.readAll();
        } catch (Exception e) {
            throw new RuntimeException("Error reading from file log: " + e.getMessage(), e);
        }
    }

    @Override
    public void clear() {
        log.clear();
    }

    @Override
    public String getDestinationName() {
        return "FILE";
    }
}

File: python-gdb-activities-20/activity20/gdb/logging/file_log_destination.py
# gdb/logging/file_log_destination.py
import pickle
import os
from typing import List
from gdb.domain.transaction import Transaction
from gdb.logging.log_destination import LogDestination

class FileLogDestination(LogDestination):
    """File storage destination implementor."""

    def __init__(self, filename: str = "transactions.log") -> None:
        self._filename = filename

    def write(self, transaction: Transaction) -> None:
        txns = self.read_all()
        txns.append(transaction)
        with open(self._filename, "wb") as f:
            pickle.dump(txns, f)

    def read_all(self) -> List[Transaction]:
        if os.path.exists(self._filename):
            try:
                with open(self._filename, "rb") as f:
                    return pickle.load(f)
            except Exception:
                return []
        return []

    def get_destination_name(self) -> str:
        return f"File: {self._filename}"

    def clear(self) -> None:
        if os.path.exists(self._filename):
            try:
                os.remove(self._filename)
            except OSError:
                pass
